File: aqp/data/mcp/tools/discovery.py
"""DataMCP tools backing the active discovery surface (phase 1).

Exposes ``data.discovery.browse``, ``data.discovery.describe``, and
``data.discovery.promote`` so agents can inventory uningested
external sources without bypassing the
:class:`DataMCPTool` boundary (AGENTS rule 22).
"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from aqp.data.discovery import DiscoveryService
from aqp.data.discovery.types import PromoteRequest
from aqp.data.mcp.base import DataMCPTool, MCPToolContext, MCPToolResult
from aqp.data.mcp.registry import register_data_mcp_tool

_service = DiscoveryService()


class BrowseDiscoveryInput(BaseModel):
    lifecycle: str | None = Field(
        default=None,
        description="Filter by lifecycle: ingested | pending | orphan | external_only.",
    )
    provider: str | None = Field(default=None)
    kind: str | None = Field(
        default=None,
        description="Filter by dataset kind (iceberg / api / external / ...).",
    )
    search: str | None = Field(default=None, description="Substring match on name / provider / description.")
    limit: int = Field(default=50, ge=1, le=500)


@register_data_mcp_tool
class BrowseDiscoveryTool(DataMCPTool):
    name = "data.discovery.browse"
    description = (
        "Browse the unified discovery surface — ingested datasets, pending external sources, "
        "Iceberg orphans, and Airbyte connection inventory in one stream. Use this before "
        "deciding whether to promote a source via the Airbyte builder."
    )
    args_schema = BrowseDiscoveryInput
    category = "discovery"
    tags = ("discovery", "catalog", "browse")
    required_scopes = ("data:read",)

    def run(
        self,
        *,
        ctx: MCPToolContext,
        lifecycle: str | None = None,
        provider: str | None = None,
        kind: str | None = None,
        search: str | None = None,
        limit: int = 50,
    ) -> MCPToolResult:
        page = _service.list(
            lifecycle=lifecycle,  # type: ignore[arg-type]
            provider=provider,
            kind=kind,
            search=search,
            limit=int(limit),
        )
        # Tenancy filter: drop entries whose Postgres row belongs to a
        # different workspace. Virtual entries (orphan: / library: /
        # airbyte:) never carry a workspace and are always shown so the
        # operator can still see uningested external sources.
        ws = ctx.workspace_id
        items: list[Any] = []
        for entry in page.items:
            entry_ws = (entry.workspace_id or "") if hasattr(entry, "workspace_id") else ""
            is_virtual = entry.id and entry.id.startswith(
                ("orphan:", "library:", "airbyte:")
            )
            if ws and entry_ws and entry_ws != ws and not is_virtual:
                # Cross-tenant data — skip it.
                continue
            items.append(entry.model_dump(mode="json"))
        return MCPToolResult(
            ok=True,
            data={
                "items": items,
                "by_lifecycle": page.by_lifecycle,
                "total": page.total,
                "next_cursor": page.next_cursor,
            },
            rows_returned=len(items),
            summary=f"discovery browse returned {len(items)} entries",
        )


class DescribeDiscoveryInput(BaseModel):
    entry_id: str = Field(..., description="Discovery entry id (DatasetCatalog id or virtual sentinel)")


@register_data_mcp_tool
class DescribeDiscoveryTool(DataMCPTool):
    name = "data.discovery.describe"
    description = (
        "Return the full DiscoveryEntry payload (description, tags, suggested connector, "
        "external spec) for a single id."
    )
    args_schema = DescribeDiscoveryInput
    category = "discovery"
    tags = ("discovery", "describe")
    required_scopes = ("data:read",)

    def run(self, *, ctx: MCPToolContext, entry_id: str) -> MCPToolResult:
        entry = _service.get(entry_id)
        if entry is None:
            return MCPToolResult(
                ok=False,
                error=f"discovery entry {entry_id!r} not found",
                summary="describe miss",
            )
        return MCPToolResult(
            ok=True,
            data=entry.model_dump(mode="json"),
            summary=f"described {entry_id}",
        )


class PromoteDiscoveryInput(BaseModel):
    entry_id: str = Field(..., description="Discovery entry id to promote")
    target_kind: str = Field(
        default="airbyte_builder",
        description="Promotion target: airbyte_builder | fetcher_stub.",
    )
    notes: str | None = Field(default=None)


@register_data_mcp_tool
class PromoteDiscoveryTool(DataMCPTool):
    name = "data.discovery.promote"
    description = (
        "Promote an uningested discovery entry into an ingestion path. "
        "Returns the deep-link URL the operator opens in the Airbyte builder."
    )
    args_schema = PromoteDiscoveryInput
    category = "discovery"
    tags = ("discovery", "promote")
    mutates = True
    required_scopes = ("data:write",)

    def run(
        self,
        *,
        ctx: MCPToolContext,
        entry_id: str,
        target_kind: str = "airbyte_builder",
        notes: str | None = None,
    ) -> MCPToolResult:
        try:
            result = _service.promote(
                entry_id,
                target_kind=target_kind,
                notes=notes,
                actor=ctx.actor,
                workspace_id=ctx.workspace_id,
                project_id=ctx.project_id,
            )
        except LookupError as exc:
            return MCPToolResult(ok=False, error=str(exc), summary="promote miss")
        return MCPToolResult(
            ok=True,
            data=result,
            summary=f"promoted {entry_id} -> {target_kind}",
            metadata={"redirect_url": result["redirect_url"]},
        )


__all__ = ["BrowseDiscoveryTool", "DescribeDiscoveryTool", "PromoteDiscoveryTool"]

File: frontend/src/lib/api/discovery.ts
import { apiFetch } from "./client";

export type DiscoveryLifecycleState =
  | "ingested"
  | "pending"
  | "orphan"
  | "external_only";

export interface DiscoveryEntry {
  id: string;
  name: string;
  provider: string;
  domain?: string | null;
  lifecycle_state: DiscoveryLifecycleState;
  dataset_kind?: string | null;
  is_ingested: boolean;
  iceberg_identifier?: string | null;
  namespace?: string | null;
  medallion_layer?: string | null;
  description?: string | null;
  docs_url?: string | null;
  source_uri?: string | null;
  tags: string[];
  spec_hash?: string | null;
  external_spec: Record<string, unknown>;
  business_metadata: Record<string, unknown>;
  data_contract: Record<string, unknown>;
  suggested_connector?: string | null;
  suggested_kind?: string | null;
  airbyte_connection_id?: string | null;
  promote_url?: string | null;
  updated_at?: string | null;
}

export interface DiscoveryPage {
  items: DiscoveryEntry[];
  total: number;
  next_cursor: number | null;
  by_lifecycle: Record<string, number>;
}

export interface CreateExternalEntryPayload {
  name: string;
  provider?: string;
  domain?: string;
  description?: string | null;
  source_uri?: string | null;
  docs_url?: string | null;
  suggested_connector?: string | null;
  suggested_kind?: string | null;
  tags?: string[];
}

export interface UpdateEntryPayload {
  description?: string | null;
  docs_url?: string | null;
  source_uri?: string | null;
  suggested_connector?: string | null;
  suggested_kind?: string | null;
  tags?: string[] | null;
  business_metadata?: Record<string, unknown> | null;
  data_contract?: Record<string, unknown> | null;
}

export interface PromoteResponse {
  entry_id: string;
  target_kind: string;
  redirect_url: string;
  builder_state: Record<string, unknown>;
}

export const DiscoveryApi = {
  list: async (
    args: {
      lifecycle?: DiscoveryLifecycleState;
      provider?: string;
      kind?: string;
      search?: string;
      cursor?: number;
      limit?: number;
    } = {},
  ): Promise<DiscoveryPage> => {
    const query: Record<string, string | number> = {
      cursor: args.cursor ?? 0,
      limit: args.limit ?? 100,
    };
    if (args.lifecycle) query.lifecycle = args.lifecycle;
    if (args.provider) query.provider = args.provider;
    if (args.kind) query.kind = args.kind;
    if (args.search) query.search = args.search;
    return apiFetch<DiscoveryPage>("/discovery/entries", { query });
  },

  describe: async (id: string): Promise<DiscoveryEntry> =>
    apiFetch<DiscoveryEntry>(`/discovery/entries/${encodeURIComponent(id)}`),

  create: async (payload: CreateExternalEntryPayload): Promise<DiscoveryEntry> =>
    apiFetch<DiscoveryEntry>("/discovery/entries", {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  patch: async (id: string, payload: UpdateEntryPayload): Promise<DiscoveryEntry> =>
    apiFetch<DiscoveryEntry>(`/discovery/entries/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    }),

  remove: async (id: string): Promise<unknown> =>
    apiFetch<unknown>(`/discovery/entries/${encodeURIComponent(id)}`, {
      method: "DELETE",
    }),

  promote: async (
    id: string,
    payload: { target_kind?: "airbyte_builder" | "fetcher_stub"; notes?: string } = {},
  ): Promise<PromoteResponse> =>
    apiFetch<PromoteResponse>(`/discovery/entries/${encodeURIComponent(id)}/promote`, {
      method: "POST",
      body: JSON.stringify({
        target_kind: payload.target_kind ?? "airbyte_builder",
        notes: payload.notes ?? null,
      }),
    }),
};
